### PR TITLE
assignment 1 solution

### PR DESCRIPTION
**Pull request by suncerock**

### CODE REVIEW

File: src/comb_filter.rs [hunk 1]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
 
     pub fn reset(&mut self) {
-        todo!("implement")
+        self.gain = 0.0;
+        self.max_delay_secs = 0.0;
     }
 
     pub fn process(&mut self, input: &[&[f32]], output: &mut [&mut [f32]]) {
-        todo!("implement");
+        let mut delay_line = vec![vec![0.0 as f32; self.num_channels]; self.max_delay_samples];
```
Comment (emurray2) [inline src/comb_filter.rs:46]: You might want to profile this and check if it's allocating memory. If so, it may be better to have the delay line as a member of this `CombFilter` and allocate it once in the `new()` method.

File: src/comb_filter.rs [hunk 2]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
+            sample_rate_hz: sample_rate_hz,
+            max_delay_samples: (max_delay_secs * sample_rate_hz) as usize,
+            num_channels: num_channels
+        }
     }
 
     pub fn reset(&mut self) {
-        todo!("implement")
+        self.gain = 0.0;
+        self.max_delay_secs = 0.0;
```
Comment (emurray2) [inline src/comb_filter.rs:42]: You may want to also recalculate `self.max_delay_samples` here

File: src/comb_filter.rs [hunk 3]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
     }
 
     pub fn process(&mut self, input: &[&[f32]], output: &mut [&mut [f32]]) {
-        todo!("implement");
+        let mut delay_line = vec![vec![0.0 as f32; self.num_channels]; self.max_delay_samples];
+        for i in 0..input.len() {
+            for channel in 0..self.num_channels {
+                output[i][channel] = input[i][channel] + self.gain * delay_line[self.max_delay_samples - 1][channel];
```
Comment (emurray2) [inline src/comb_filter.rs:49]: Have you tested this with `self.max_delay_samples` equal to zero? You might want to add a test for this or add a test for `reset()`.

File: src/comb_filter.rs [hunk 4]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
+                for delay_index in (1..self.max_delay_samples).rev() {
+                    delay_line[delay_index][channel] = delay_line[delay_index - 1][channel];
+                }
+                delay_line[0][channel] = match self.filter_type {
```
Comment (emurray2) [inline src/comb_filter.rs:53]: Might want to add a try or something at the beginning of `process()` to check if the delay line is empty or `max_delay_samples` is zero.

File: src/comb_filter.rs [hunk 5]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
+                }
+            }
+        }
     }
 
     pub fn set_param(&mut self, param: FilterParam, value: f32) -> Result<(), Error> {
-        todo!("implement")
+        match param {
+            FilterParam::Gain => self.set_gain(value),
+            FilterParam::Delay => self.set_delay(value),
```
Comment (emurray2) [inline src/comb_filter.rs:64]: In case the user expected to be able to set delay to a higher value, I would print a warning in the console when running in debug.

File: src/comb_filter.rs [hunk 6]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
     }
 
     // TODO: feel free to define other functions for your own use
+    fn set_gain(&mut self, value: f32) -> Result<(), Error> {
+        self.gain = value;
+        Ok(())
+    }
+
+    fn set_delay(&mut self, value: f32) -> Result<(), Error> {
+        self.max_delay_secs = value;
```
Comment (emurray2) [inline src/comb_filter.rs:82]: It would be a good idea to recalculate `self.max_delay_samples` here and check if there is enough memory to allocate this amount of samples.

File: src/main.rs [hunk 7]
```diff
@@ -25,11 +39,248 @@ fn main() {
...
+        }
+        for t in 80..160 {
+            assert!((output[t][0] - 0.75).abs() < 1.0e-7);
+        }
+        for t in 160..240 {
+            assert!((output[t][0] - 0.875).abs() < 1.0e-7);
+        }
+    }
+
+}
```
Comment (emurray2) [inline src/main.rs:286]: I would consider looking into implementing the blocks which I'm working on doing myself and adding some tests for `reset()`.

**Review by emurray2 — COMMENTED** [review]
I would consider adding some more tests for changing and setting parameters to make sure they change the result you get and also handle some edge cases. Other than that, everything looks great. Awesome job 🙌 !

File: README.md
Comment (YellowWaveSr) [inline README.md:1]: How about uploading the python files as well? Besides, maybe you can have one more audio file for testing.

File: fir_comparison.jpg
Comment (YellowWaveSr) [inline fir_comparison.jpg:1]: Could you please also provide the plots of the waveform before and after the filter?

File: src/comb_filter.rs [hunk 8]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
 
 impl CombFilter {
     pub fn new(filter_type: FilterType, max_delay_secs: f32, sample_rate_hz: f32, num_channels: usize) -> Self {
```
Comment (YellowWaveSr) [inline src/comb_filter.rs:29]: Is it a better idea if the delay buffer is defined here?

File: src/comb_filter.rs [hunk 9]
```diff
@@ -21,26 +27,62 @@ pub enum Error {
...
     }
 
     pub fn reset(&mut self) {
-        todo!("implement")
+        self.gain = 0.0;
+        self.max_delay_secs = 0.0;
     }
 
     pub fn process(&mut self, input: &[&[f32]], output: &mut [&mut [f32]]) {
-        todo!("implement");
```
Comment (YellowWaveSr) [inline src/comb_filter.rs:32]: So far I could only find the process method but I cannot find the details about how to deal with different block sizes. Maybe you could have one more method to do processing for each block and another method to make the whole signal into blocks.

File: src/main.rs [hunk 10]
```diff
@@ -25,11 +39,248 @@ fn main() {
...
+    }).unwrap();
+
+    for sample in output.iter() {
+        for &value in sample.iter() {
+            writer.write_sample((value * i16::MAX as f32) as i16).unwrap();
+        }
     }
 }
+
+#[cfg(test)]
```
Comment (YellowWaveSr) [inline src/main.rs:75]: I like your test methods. This is really convenient and neat.
